File: src/main/java/net/atos/zac/util/OpenZaakReadinessHealthCheck.java
/*
 * SPDX-FileCopyrightText: 2021 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.util;

import java.time.LocalDateTime;

import io.opentelemetry.instrumentation.annotations.WithSpan;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import org.eclipse.microprofile.health.HealthCheck;
import org.eclipse.microprofile.health.HealthCheckResponse;
import org.eclipse.microprofile.health.Readiness;

import io.opentelemetry.instrumentation.annotations.WithSpan;
import nl.info.client.zgw.ztc.ZtcClientService;
import nl.info.client.zgw.ztc.model.CatalogusListParameters;
import nl.info.zac.configuratie.ConfiguratieService;

@Readiness
@ApplicationScoped
public class OpenZaakReadinessHealthCheck implements HealthCheck {

    private static final CatalogusListParameters CATALOGUS_LIST_PARAMETERS = new CatalogusListParameters();

    private final ZtcClientService ztcClientService;

    @Inject
    OpenZaakReadinessHealthCheck(ZtcClientService ztcClientService, ConfiguratieService configuratieService) {
        this.ztcClientService = ztcClientService;

        CATALOGUS_LIST_PARAMETERS.setDomein(configuratieService.readCatalogusDomein());
    }

    @Override
    @WithSpan(value = "GET OpenZaakReadinessHealthCheck")
    public HealthCheckResponse call() {
        try {
            ztcClientService.listCatalogus(CATALOGUS_LIST_PARAMETERS);
            return HealthCheckResponse.up(OpenZaakReadinessHealthCheck.class.getName());
        } catch (final Exception exception) {
            return HealthCheckResponse.named(OpenZaakReadinessHealthCheck.class.getName())
                    .withData("time", LocalDateTime.now().toString())
                    .withData("error", exception.getMessage())
                    .down()
                    .build();
        }
    }
}

File: src/main/kotlin/nl/info/zac/health/OpenZaakReadinessHealthCheck.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2025 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.health

import io.opentelemetry.instrumentation.annotations.WithSpan
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import nl.info.client.zgw.ztc.ZtcClientService
import nl.info.client.zgw.ztc.model.CatalogusListParameters
import nl.info.zac.configuratie.ConfigurationService
import nl.info.zac.util.AllOpen
import org.eclipse.microprofile.health.HealthCheck
import org.eclipse.microprofile.health.HealthCheckResponse
import org.eclipse.microprofile.health.Readiness
import java.time.LocalDateTime

@Readiness
@ApplicationScoped
@AllOpen
class OpenZaakReadinessHealthCheck @Inject constructor(
    private val ztcClientService: ZtcClientService,
    private val configurationService: ConfigurationService
) : HealthCheck {

    @WithSpan(value = "GET OpenZaakReadinessHealthCheck")
    override fun call(): HealthCheckResponse =
        try {
            configurationService.readCatalogusDomein()
                .let { domein -> CatalogusListParameters().also { it.domein = domein } }
                .let(ztcClientService::listCatalogus)
            HealthCheckResponse.up(OpenZaakReadinessHealthCheck::class.java.name)
        } catch (@Suppress("TooGenericExceptionCaught") exception: RuntimeException) {
            HealthCheckResponse.named(OpenZaakReadinessHealthCheck::class.java.name)
                .withData("time", LocalDateTime.now().toString())
                .withData("error", exception.message)
                .down()
                .build()
        }
}
